### PR TITLE
Iter5

### PR DESCRIPTION
**Pull request by sur1k1**

### CODE REVIEW

File: internal/agent/config/flags.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+type Config struct {
+	Address 				string `env:"ADDRESS"`
+	PollInterval 		int64 `env:"POLL_INTERVAL"`
+	ReportInterval	int64 `env:"REPORT_INTERVAL"`
+}
+
+func Setup() (*AgentOptions, error) {
+	var opts AgentOptions
```
Comment (arttet) [inline internal/agent/config/flags.go:24]: [SUGGESTION] If I were you, I would use `Builder` pattern.

File: internal/agent/config/flags.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+)
+
+type AgentOptions struct {
+	AddressServer  string
+	PollInterval   time.Duration
+	ReportInterval time.Duration
+}
+
+type Config struct {
+	Address 				string `env:"ADDRESS"`
```
Comment (arttet) [inline internal/agent/config/flags.go:18]: [CHANGE REQUEST] Please format your code.

File: internal/agent/metric/metric.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+func NewMetricStorage() *MetricStorage {
+	return &MetricStorage{
+		GaugeMap: map[string]float64{},
+		CounterMap: map[string]int64{},
+	}
+}
+
+// Описание интерфейса метрик
+type Metric interface {
+	MetricUpdaterIntf
```
Comment (arttet) [inline internal/agent/metric/metric.go:37]: [CLARIFICATION] What is it?

File: internal/agent/metric/metric.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+// Интерфейс обновления метрик
+type MetricUpdaterIntf interface {
+	UpdateCounter()
+	UpdateGauge()
+}
+
+// Получение обновление по pollInterval
+func MetricUpdater(mu MetricUpdaterIntf, flagOpts *config.AgentOptions) {
+	for {
```
Comment (arttet) [inline internal/agent/metric/metric.go:49]: [CHANGE REQUEST] Please don't use infinite loops

File: internal/server/handlers/metric_list.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import (
+	"fmt"
+	"net/http"
+)
+
+type AllMetricsGetter interface {
+	GetAllMetrics() map[string]string
+}
+
+func MetricListHandler(s AllMetricsGetter) http.HandlerFunc {
```
Comment (arttet) [inline internal/server/handlers/metric_list.go:12]: [CHANGE REQUEST] It should be a method.

File: internal/server/handlers/update_handler.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
+package handlers
+
+import (
+	"net/http"
+
+	"github.com/go-chi/chi/v5"
+)
+
+const (
+	Gauge = string("gauge")
```
Comment (arttet) [inline internal/server/handlers/update_handler.go:10]:
```suggestion
	Gauge = "gauge"
```

Comment (arttet) [inline internal/server/handlers/update_handler.go:10]:
```suggestion
	GaugeTypeStr = string("gauge")
```

Comment (arttet) [reply internal/server/handlers/update_handler.go:10]: [CHANGE REQUEST] Please don't repeat yourself.

**Review by arttet — CHANGES_REQUESTED** [review]
- [ ] Split the application into layers
https://github.com/golang-standards/project-layout
https://medium.com/@deanrubin/the-three-layered-architecture-fe30cb0e4a6#:~:text=The%20presentation%20layer%20is%20the,Update%20or%20Delete)%20operation%20flow.
- [ ] Please integrate CodeCov https://about.codecov.io/blog/getting-started-with-code-coverage-for-golang/
- [ ] Please read and use the Conventional Commits specification https://www.conventionalcommits.org/en/v1.0.0/
- [ ] Please integrate the linter https://github.com/golangci/golangci-lint-action
- [ ] Please delete deprecated constraints in CI
https://github.com/sur1k1/go-metrics/blob/c3169a397a44c51f85121c37fe008388539e1770/.github/workflows/mertricstest.yml#L67

File: internal/agent/metric/metric.go (hunk 4 above)
Comment (sur1k1) [reply internal/agent/metric/metric.go:49]: What to use instead?

File: .github/workflows/mertricstest.yml [hunk 7]
```diff
@@ -64,7 +64,6 @@ jobs:
 
       - name: "Code increment #1"
         if: |
```
Comment (arttet) [inline .github/workflows/mertricstest.yml:66]: [CHANGE REQUEST] This test won't run in `main`. But we have to run it there.

File: cmd/agent/main.go [hunk 8]
```diff
@@ -1,3 +1,42 @@
...
+
+	"github.com/sur1k1/go-metrics/internal/agent/config"
+	"github.com/sur1k1/go-metrics/internal/agent/metric"
+)
+
+func main() {
+	gracefulShutdown()
+	flagOpts, err := config.Setup()
+	if err != nil{
+		panic(err)
```
Comment (arttet) [inline cmd/agent/main.go:18]: [CHANGE REQUEST] Please don't panic.
https://github.com/uber-go/guide/blob/master/style.md#dont-panic

File: cmd/agent/main.go [hunk 9]
```diff
@@ -1,3 +1,42 @@
...
+	wg.Wait()
+}
+
+func gracefulShutdown() {
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+	go func() {
+		<-c
+		log.Println("Shutting down gracefully")
+		os.Exit(0)
```
Comment (arttet) [inline cmd/agent/main.go:40]: [SUGGESTION] This is not a graceful shutdown.

File: internal/agent/metric/metric.go [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+	"math/rand/v2"
+	"net/http"
+	"runtime"
+	"time"
+
+	"github.com/sur1k1/go-metrics/internal/agent/config"
+)
+
+const (
+	reportForm			string = "http://%s/update/%s/%s/%v"
```
Comment (arttet) [inline internal/agent/metric/metric.go:16]: [SUGGESTION] Please format your code

File: internal/agent/metric/metric.go [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+
+// Интерфейс отправки метрик
+type MetricSenderIntf interface {
+	Send(client http.Client, addr string) error
+}
+
+// Отправка метрик на сервер по reportInterval
+func MetricSender(ms MetricSenderIntf, flagOpts *config.AgentOptions) {
+	var client http.Client
+	for {
```
Comment (arttet) [inline internal/agent/metric/metric.go:98]: [SUGGESTION] Please don't use infinite loops.

File: internal/agent/metric/metric.go [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+		}
+		defer resp.Body.Close()
+		
+		_, err = io.ReadAll(resp.Body)
+		if err != nil{
+			return err
+		}
+	}
+
+	for name, value := range m.CounterMap{
```
Comment (arttet) [inline internal/agent/metric/metric.go:131]: [SUGGESTION] This function has a huge Cognitive Complexity.

File: internal/server/rest/update_handler.go [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+func NewUpdateHandler(router *chi.Mux, svc UpdateService) {
+	handler := &UpdateHandler{
+		Service: svc,
+	}
+
+	router.Post("/update/{type}/{metric}/{value}", handler.Update())
+}
+
+
+func (s *UpdateHandler) Update() http.HandlerFunc {
```
Comment (arttet) [inline internal/server/rest/update_handler.go:32]:
```suggestion
func (s *UpdateHandler) Update(w http.ResponseWriter, r *http.Request) {
```

File: service/service.go [hunk 14]
```diff
@@ -0,0 +1,49 @@
+package service
+
+type ServiceRepository interface {
```
Comment (arttet) [inline service/service.go:3]: [CLARIFICATION] Why is it here?

**Review by arttet — CHANGES_REQUESTED** [review]

File: service/service.go (hunk 14 above)
Comment (sur1k1) [reply service/service.go:3]: https://medium.com/@deanrubin/the-three-layered-architecture-fe30cb0e4a6#:~:text=The%20presentation%20layer%20is%20the,Update%20or%20Delete)%20operation%20flow

Comment (sur1k1) [reply service/service.go:3]: https://github.com/bxcodec/go-clean-arch/blob/master/internal/rest/article.go

**Pull request merged**